### ISSUE TITLE
Active Directory Integrated authentication broken when used with newer version of Microsoft.Identity.Client

### ISSUE THREAD

**Issue by daniel-chambers** [dotnet/SqlClient#744]
### Describe the bug
If `Microsoft.Identity.Client` 4.16.0 or higher is used with this package, `Active Directory Integrated` authentication to a SQL Managed Instance in Azure fails.

```
Unhandled exception. System.AggregateException: One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error: )))
 ---> System.AggregateException: One or more errors occurred. (One or more errors occurred. (Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error: ))
 ---> System.AggregateException: One or more errors occurred. (Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error: )
 ---> MSAL.NetCore.4.16.0.0.MsalClientException:
        ErrorCode: parsing_wstrust_response_failed
Microsoft.Identity.Client.MsalClientException: Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error:
 ---> MSAL.NetCore.4.16.0.0.MsalServiceException:
        ErrorCode: federated_service_returned_error
Microsoft.Identity.Client.MsalServiceException: Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error:
   at Microsoft.Identity.Client.WsTrust.WsTrustWebRequestManager.GetWsTrustResponseAsync(WsTrustEndpoint wsTrustEndpoint, String wsTrustRequest, RequestContext requestContext)
   at Microsoft.Identity.Client.WsTrust.CommonNonInteractiveHandler.GetWsTrustResponseAsync(UserAuthType userAuthType, String cloudAudienceUrn, WsTrustEndpoint endpoint, String username, SecureString securePassword)
        StatusCode: 401
        ResponseBody:
        Headers: Cache-Control: no-store, no-cache
Pragma: no-cache
Vary: Origin
X-Content-Type-Options: nosniff
Access-Control-Allow-Origin: https://login.microsoftonline.com
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
x-ms-request-id: f3729d65-f414-43d3-8900-ab8e9be0d100
x-ms-ests-server: 2.1.11063.14 - HKG2 ProdSlices
Set-Cookie: fpc=****REDACTED****; expires=Fri, 30-Oct-2020 03:26:16 GMT; path=/; secure; HttpOnly; SameSite=None, x-ms-gateway-slice=prod; path=/; secure; httponly, stsservicecookie=ests; path=/; secure; samesite=none; httponly
WWW-Authenticate: Negotiate
Date: Wed, 30 Sep 2020 03:26:16 GMT

   --- End of inner exception stack trace ---
   at Microsoft.Identity.Client.WsTrust.CommonNonInteractiveHandler.GetWsTrustResponseAsync(UserAuthType userAuthType, String cloudAudienceUrn, WsTrustEndpoint endpoint, String username, SecureString securePassword)
   at Microsoft.Identity.Client.WsTrust.CommonNonInteractiveHandler.PerformWsTrustMexExchangeAsync(String federationMetadataUrl, String cloudAudienceUrn, UserAuthType userAuthType, String username, SecureString password)
   at Microsoft.Identity.Client.Internal.Requests.IntegratedWindowsAuthRequest.FetchAssertionFromWsTrustAsync()
   at Microsoft.Identity.Client.Internal.Requests.IntegratedWindowsAuthRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenByIntegratedWindowsAuthParameters integratedWindowsAuthParameters, CancellationToken cancellationToken)
Inner Excception: MSAL.NetCore.4.16.0.0.MsalServiceException:
        ErrorCode: federated_service_returned_error
Microsoft.Identity.Client.MsalServiceException: Federated service at https://autologon.microsoftazuread-sso.com/****REDACTED****/winauth/trust/2005/windowstransport?client-request-id=5a3440cf-20cb-4b6f-b4c5-cf3e7bdd1fa5 returned error:
   at Microsoft.Identity.Client.WsTrust.WsTrustWebRequestManager.GetWsTrustResponseAsync(WsTrustEndpoint wsTrustEndpoint, String wsTrustRequest, RequestContext requestContext)
   at Microsoft.Identity.Client.WsTrust.CommonNonInteractiveHandler.GetWsTrustResponseAsync(UserAuthType userAuthType, String cloudAudienceUrn, WsTrustEndpoint endpoint, String username, SecureString securePassword)
        StatusCode: 401
        ResponseBody:
        Headers: Cache-Control: no-store, no-cache
Pragma: no-cache
Vary: Origin
X-Content-Type-Options: nosniff
Access-Control-Allow-Origin: https://login.microsoftonline.com
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
x-ms-request-id: f3729d65-f414-43d3-8900-ab8e9be0d100
x-ms-ests-server: 2.1.11063.14 - HKG2 ProdSlices
Set-Cookie: fpc=****REDACTED****; expires=Fri, 30-Oct-2020 03:26:16 GMT; path=/; secure; HttpOnly; SameSite=None, x-ms-gateway-slice=prod; path=/; secure; httponly, stsservicecookie=ests; path=/; secure; samesite=none; httponly
WWW-Authenticate: Negotiate
Date: Wed, 30 Sep 2020 03:26:16 GMT

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Microsoft.Data.SqlClient.ActiveDirectoryAuthenticationProvider.<>c__DisplayClass3_0.<<AcquireTokenAsync>b__0>d.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.<>c__DisplayClass144_1.<GetFedAuthToken>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   --- End of inner exception stack trace ---
   at Microsoft.Data.ProviderBase.DbConnectionPool.CheckPoolBlockingPeriod(Exception e)
   at Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open()
   at Program.main(String[] argv) in C:\Users\dchambers\Code\Spiking\SqlAadIntegratedAuth\Program.fs:line 10
```

### To reproduce
Create a .NET Core 3.1 console app with these PackageReferences:

```
<PackageReference Include="Microsoft.Data.SqlClient" Version="2.0.1" />
<PackageReference Include="Microsoft.Identity.Client" Version="4.16.0" />
```

And in your `Program.fs` file:

```fsharp
open System
open Microsoft.Data.SqlClient

[<EntryPoint>]
let main argv =
    use sqlConnection = new SqlConnection("Data Source=yourownsqlmanagedinstance.database.windows.net;Initial Catalog=YourDb;Authentication=Active Directory Integrated;")
    sqlConnection.Open()
    0 // return an integer exit code
```

If you downgrade `Microsoft.Identity.Client` to 4.15.0, the problem goes away.

### Expected behavior
A database connection to successfully open. 

### Further technical details
Microsoft.Data.SqlClient version: 2.0.1
.NET target: .NET Core 3.1
SQL Server version: SQL Managed Instance in Azure
Operating system: Windows 10 1909

**Additional context**
This bites users of the Paket dependency manager straight away, because Paket respects the version bounds defined on `Microsoft.Data.SqlClient` and by default automatically tries to upgrade `Microsoft.Identity.Client` to the latest version because the version bound is `>= 4.14.0`, which causes this error. I tested 4.14.0 and 4.15.0 and both these versions work fine. 4.16.0 is where it stops working.

**Comment by cheenamalhotra** [comment]
Hi @daniel-chambers 

The higher version of Microsoft.Identity.Client where this issue occurs must be reported to MSAL team in their repository (https://github.com/AzureAD/microsoft-authentication-library-for-dotnet).

Moreover you should be able to reproduce the issue by using just the MSAL library and trying to acquire access token for Azure Integrated authentication by mocking driver's implementation in a standalone app.

https://github.com/dotnet/SqlClient/blob/d14c0b1196de6271b36251d7aff63cfeadcad799/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs#L156

Since we don't directly depend on the corrupted version, and v4.16.0 is expected to be backwards compatible with v4.14.0 as per semantics, it falls out of scope for us to be able to provide a fix in this case.

**Comment by cheenamalhotra** [comment]
Hi @daniel-chambers 

We'll close the issue as it comes from external component.

**Comment by daniel-chambers** [comment]
For others coming here: this appears to be have been fixed in MSAL 4.20.0 (https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1988)
